core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -120,29 +125,42 @@ private[server] class MetadataCache(brokerId: Int) extends Logging {
     }
   }
 
-  def updateCache(updateMetadataRequest: UpdateMetadataRequest,
+  def updateCache(correlationId: Int, updateMetadataRequest: UpdateMetadataRequest,
                   brokerId: Int,
                   stateChangeLogger: StateChangeLogger) {
     inWriteLock(partitionMetadataLock) {
-      aliveBrokers = updateMetadataRequest.aliveBrokers.map(b => (b.id, b)).toMap
-      updateMetadataRequest.partitionStateInfos.foreach { case(tp, info) =>
-        if (info.leaderIsrAndControllerEpoch.leaderAndIsr.leader == LeaderAndIsr.LeaderDuringDelete) {
+      aliveBrokers = updateMetadataRequest.liveBrokers.asScala.map { broker =>
+        val endPoints = broker.endPoints.asScala.map { case (protocol, ep) =>
```

Comment:
Maybe you could use `mapValues` instead of `map` here? Since you call `toMap` at the end, the fact that `mapValues` is lazy is actually a benefit instead of a potential issue.